android/gradle_project/allegro/src/main/java/org/liballeg/android/AllegroDialog.java

```diff
@@ -0,0 +1,658 @@
+        builder.setCancelable(true);
+
+        if (0 != (flags & AllegroDialogConst.ALLEGRO_MESSAGEBOX_WARN))
+            builder.setIcon(android.R.drawable.ic_dialog_alert);
+        else if (0 != (flags & AllegroDialogConst.ALLEGRO_MESSAGEBOX_ERROR))
+            builder.setIcon(android.R.drawable.ic_dialog_alert); // ic_delete
+        else if (0 != (flags & AllegroDialogConst.ALLEGRO_MESSAGEBOX_QUESTION))
+            builder.setIcon(android.R.drawable.ic_dialog_info);
+
+        builder.setOnCancelListener(new DialogInterface.OnCancelListener() {
+            @Override
+            public void onCancel(DialogInterface dialog) {
+                dialog.dismiss();
+            }
+        });
+
+        builder.setOnDismissListener(new DialogInterface.OnDismissListener() {
+            @Override
+            public void onDismiss(DialogInterface dialog) {
+                if (dialog == currentDialog)
+                    currentDialog = null;
+
+                interrupter.interrupt();
+            }
+        });
+
+        // configure the buttons
+        boolean wantCustomButtons = !buttons.equals("");
+        String[] buttonText = wantCustomButtons ? buttons.split("\\|") : null;
+
+        if (!wantCustomButtons) {
+            builder.setPositiveButton(android.R.string.ok, resultSetter.generate(POSITIVE_BUTTON));
+
+            if (0 != (flags & (AllegroDialogConst.ALLEGRO_MESSAGEBOX_OK_CANCEL | AllegroDialogConst.ALLEGRO_MESSAGEBOX_YES_NO))) {
+                // unfortunately, android.R.string.yes and android.R.string.no
+                // are deprecated and resolve to android.R.string.ok and
+                // android.R.string.cancel, respectively.
+                builder.setNegativeButton(android.R.string.cancel, resultSetter.generate(NEGATIVE_BUTTON));
+            }
+        }
+        else if (buttonText.length == 1) {
+            builder.setPositiveButton(buttonText[0], resultSetter.generate(POSITIVE_BUTTON));
+        }
+        else if (buttonText.length == 2) {
+            builder.setPositiveButton(buttonText[0], resultSetter.generate(POSITIVE_BUTTON));
+            builder.setNegativeButton(buttonText[1], resultSetter.generate(NEGATIVE_BUTTON));
+        }
+        else if (buttonText.length >= 3) { // we support up to 3 buttons
+            builder.setPositiveButton(buttonText[0], resultSetter.generate(POSITIVE_BUTTON));
+            builder.setNegativeButton(buttonText[1], resultSetter.generate(NEGATIVE_BUTTON));
+            builder.setNeutralButton(buttonText[2], resultSetter.generate(NEUTRAL_BUTTON));
+        }
+
+        // create the alert dialog
+        return builder.create();
+    }
+
+    private void showDialogWithImmersiveModeFix(Dialog dialog, Activity activity)
+    {
+        ImmersiveDialogWrapper dlg = new ImmersiveDialogWrapper(dialog);
+        dlg.show(activity);
+    }
+
+    private Looper myLooper()
+    {
+        try { Looper.prepare(); } catch (Exception e) { ; }
```

Comment:
Why do you need to discard exceptions here? Is it worth logging them?

Reply:
`Looper.prepare()` initializes the current thread as a Looper. As you can see at [Looper.java](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/os/Looper.java;drc=5ba5545833d49fefe952dc80731c396eadc3e6b5;l=102-117), the method can throw a runtime exception that says "Only one Looper may be created per thread". The class constructor does nothing special. Would you like this logged? It doesn't seem useful and we would have this message show up repeatedly on the logs.

Reply:
See also: [MessageQueue.nativeInit](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/jni/android_os_MessageQueue.cpp;l=206-215?q=android_os_MessageQueue_nativeInit&ss=android%2Fplatform%2Fsuperproject%2Fmain). Would you like this logged?

Reply:
Ok, that makes sense. Could you add a comment about this? Something like `// Looper will raise an exception if one already exists in the thread.`

Reply:
Sure